src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,171 @@
+ * for example you may wish to pool expensive state objects.</li>
+ * </ul>
+ *
+ * <h3>More technical explanation</h3>
+ *
+ * <p>{@link #updateSnapshots} should be called before any modification.
+ * This will save the state of this participant using {@link #createSnapshot} if no state was already saved for that transaction.
+ * When the transaction is aborted and changes need to be rolled back, {@link #revertToSnapshot} will be called
+ * to signal that the current state should revert to that of the snapshot.
+ * The snapshot object is then {@linkplain #releaseSnapshot released}, and can be cached for subsequent use, or discarded.
+ *
+ * <p>When an outer transaction is committed, {@link #revertToSnapshot} will not be called so that the current state of this participant
+ * is retained. {@link #onCommit} will be called after the transaction is closed
+ * and then {@link #releaseSnapshot} will be called because the snapshot is not necessary anymore.
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
+ */
+public abstract class SnapshotJournal<T> implements Transaction.CloseCallback, TransactionContext.RootCloseCallback {
+    //Neo TODO: Remove after migrations have been established. This is more for info really.
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    private static int DEEPEST_LAYER = -1;
+    @Nullable
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    private static SnapshotJournal<?> DEEPEST_SNAPSHOT = null;
+    private final ArrayList<T> snapshots = new ArrayList<>();
+
+    @Nullable
+    private T originalState = null;
+
+    /**
+     * Return a new <b>nonnull</b> object containing the current state of this participant.
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    protected abstract T createSnapshot();
+
+    /**
+     * Roll back to a state previously created by {@link #createSnapshot}.
+     */
+    protected abstract void revertToSnapshot(T snapshot);
+
+    /**
+     * Signals that the snapshot will not be used anymore, and is safe to cache for future calls to {@link #createSnapshot},
+     * or discard entirely.
+     */
+    protected void releaseSnapshot(T snapshot) {}
+
+    /**
+     * Called after an outer transaction succeeded,
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     * <p>
+     * Assume that without any implementation, data on something like BlockEntities that expect to be told when its data is written to, could be potentially lost on world save.
+     *
+     * @param originalState state of this participant before the transactional operation.
+     *                      This corresponds to the first {@link #createSnapshot() snapshot} that was created in the transactional operation.
+     */
+    protected void onCommit(T originalState) {}
```

Comment:
This should be called `onFinalCommit`/`onRootCommit` to make it clear that it only applies to the root transaction being committed.